File: src/elixir/operational_transform/test/arch_astro/operational_transform/text_operation_test.exs
defmodule ArchAstro.OperationalTransform.TextOperationTest do
  use ExUnit.Case, async: true

  alias ArchAstro.OperationalTransform.TextOperation, as: Op

  doctest ArchAstro.OperationalTransform.TextOperation

  describe "building and lengths" do
    test "tracks base and target lengths" do
      op = Op.new() |> Op.retain(3) |> Op.insert("abc") |> Op.delete(2)
      assert op.base_length == 5
      assert op.target_length == 6
    end

    test "merges consecutive components of the same kind" do
      op = Op.new() |> Op.retain(2) |> Op.retain(3) |> Op.insert("a") |> Op.insert("b")
      assert Op.to_list(op) == [5, "ab"]
    end

    test "normalizes adjacent insert/delete to insert-first" do
      a = Op.new() |> Op.retain(1) |> Op.delete(2) |> Op.insert("xy")
      b = Op.new() |> Op.retain(1) |> Op.insert("xy") |> Op.delete(2)
      assert Op.to_list(a) == [1, "xy", -2]
      assert Op.to_list(a) == Op.to_list(b)
      assert Op.apply!(a, "abc") == "axy"
    end

    test "ignores empty components" do
      op = Op.new() |> Op.retain(0) |> Op.insert("") |> Op.delete(0)
      assert Op.to_list(op) == []
      assert Op.noop?(op)
    end

    test "measures in code points, not bytes or graphemes" do
      # "👍" is 1 code point (4 bytes); "é" is 2 code points (1 grapheme).
      op = Op.new() |> Op.retain(1) |> Op.delete(2)
      assert Op.apply!(op, "👍é") == "👍"
    end
  end

  describe "wire format" do
    test "round-trips" do
      op = Op.new() |> Op.retain(2) |> Op.insert("hi 🚀") |> Op.delete(3) |> Op.retain(1)
      assert op |> Op.to_list() |> Op.from_list!() == op
    end

    test "rejects garbage" do
      assert {:error, _} = Op.from_list([1, :nope])
      assert {:error, _} = Op.from_list([0])
      assert {:error, _} = Op.from_list(%{})
    end
  end

  describe "apply" do
    test "applies a mixed operation" do
      op = Op.new() |> Op.retain(6) |> Op.insert("world") |> Op.delete(5)
      assert Op.apply!(op, "hello there") == "hello world"
    end

    test "errors on length mismatch" do
      op = Op.new() |> Op.retain(3)
      assert {:error, {:length_mismatch, _}} = Op.apply(op, "ab")
    end
  end

  describe "invert" do
    test "undoes the operation" do
      doc = "hello *world*"
      op = Op.new() |> Op.retain(6) |> Op.delete(7) |> Op.insert("there")
      applied = Op.apply!(op, doc)
      assert Op.apply!(Op.invert(op, doc), applied) == doc
    end
  end

  describe "compose" do
    test "composition equals sequential application" do
      doc = "abcdef"
      a = Op.new() |> Op.retain(3) |> Op.insert("X") |> Op.retain(3)
      b = Op.new() |> Op.delete(2) |> Op.retain(5)
      composed = Op.compose!(a, b)
      assert Op.apply!(composed, doc) == Op.apply!(b, Op.apply!(a, doc))
    end

    test "rejects incompatible lengths" do
      a = Op.new() |> Op.retain(3)
      b = Op.new() |> Op.retain(4)
      assert {:error, {:compose_length_mismatch, _}} = Op.compose(a, b)
    end
  end

  describe "transform" do
    test "TP1: both orders converge" do
      doc = "she is a girl"
      a = Op.new() |> Op.retain(9) |> Op.insert("good ") |> Op.retain(4)
      b = Op.new() |> Op.delete(3) |> Op.insert("he") |> Op.retain(10)

      {a1, b1} = Op.transform!(a, b)

      assert Op.apply!(b1, Op.apply!(a, doc)) == Op.apply!(a1, Op.apply!(b, doc))
      assert Op.apply!(b1, Op.apply!(a, doc)) == "he is a good girl"
    end

    test "insert tie-break: left operand's insert comes first" do
      doc = "ab"
      a = Op.new() |> Op.retain(1) |> Op.insert("X") |> Op.retain(1)
      b = Op.new() |> Op.retain(1) |> Op.insert("Y") |> Op.retain(1)

      {a1, b1} = Op.transform!(a, b)
      assert Op.apply!(b1, Op.apply!(a, doc)) == "aXYb"
      assert Op.apply!(a1, Op.apply!(b, doc)) == "aXYb"
    end

    test "overlapping deletes converge" do
      doc = "abcdef"
      a = Op.new() |> Op.retain(1) |> Op.delete(3) |> Op.retain(2)
      b = Op.new() |> Op.retain(2) |> Op.delete(3) |> Op.retain(1)

      {a1, b1} = Op.transform!(a, b)
      assert Op.apply!(b1, Op.apply!(a, doc)) == Op.apply!(a1, Op.apply!(b, doc))
      assert Op.apply!(b1, Op.apply!(a, doc)) == "af"
    end

    test "rejects different base lengths" do
      a = Op.new() |> Op.retain(3)
      b = Op.new() |> Op.retain(4)
      assert {:error, {:transform_length_mismatch, _}} = Op.transform(a, b)
    end
  end

  describe "transform_index" do
    test "inserts before the index shift it right" do
      op = Op.new() |> Op.insert("ab") |> Op.retain(5)
      assert Op.transform_index(3, op) == 5
    end

    test "inserts after the index leave it alone" do
      op = Op.new() |> Op.retain(4) |> Op.insert("ab") |> Op.retain(1)
      assert Op.transform_index(3, op) == 3
    end

    test "deletes spanning the index clamp it" do
      op = Op.new() |> Op.retain(1) |> Op.delete(4)
      assert Op.transform_index(3, op) == 1
    end

    test "deletes before the index shift it left" do
      op = Op.new() |> Op.delete(2) |> Op.retain(3)
      assert Op.transform_index(4, op) == 2
    end
  end
end

File: src/ts/operational_transform/test/text-operation.test.ts
import { describe, expect, test } from "vitest";
import {
  TextOperation,
  cpLength,
  utf16OffsetToCp,
  cpToUtf16Offset,
} from "../src/text-operation.js";

function op(): TextOperation {
  return new TextOperation();
}

describe("building and lengths", () => {
  test("tracks base and target lengths", () => {
    const o = op().retain(3).insert("abc").delete(2);
    expect(o.baseLength).toBe(5);
    expect(o.targetLength).toBe(6);
  });

  test("merges consecutive components of the same kind", () => {
    const o = op().retain(2).retain(3).insert("a").insert("b");
    expect(o.toJSON()).toEqual([5, "ab"]);
  });

  test("normalizes adjacent insert/delete to insert-first", () => {
    const a = op().retain(1).delete(2).insert("xy");
    const b = op().retain(1).insert("xy").delete(2);
    expect(a.toJSON()).toEqual([1, "xy", -2]);
    expect(a.toJSON()).toEqual(b.toJSON());
    expect(a.apply("abc")).toBe("axy");
  });

  test("ignores empty components", () => {
    const o = op().retain(0).insert("").delete(0);
    expect(o.toJSON()).toEqual([]);
    expect(o.isNoop()).toBe(true);
  });

  test("measures in code points, not UTF-16 units", () => {
    // "👍" is 1 code point but 2 UTF-16 units; "é" (é decomposed) is 2
    // code points but 1 grapheme.
    const o = op().retain(1).delete(2);
    expect(o.apply("\u{1F44D}e\u0301")).toBe("\u{1F44D}");
  });
});

describe("wire format", () => {
  test("round-trips", () => {
    const o = op().retain(2).insert("hi 🚀").delete(3).retain(1);
    expect(TextOperation.fromJSON(o.toJSON()).equals(o)).toBe(true);
  });

  test("matches the Elixir library's format", () => {
    const o = op().retain(6).insert("world").delete(5);
    expect(o.toJSON()).toEqual([6, "world", -5]);
    expect(o.apply("hello there")).toBe("hello world");
  });

  test("rejects garbage", () => {
    expect(() => TextOperation.fromJSON([1, null])).toThrow();
    expect(() => TextOperation.fromJSON([0])).toThrow();
    expect(() => TextOperation.fromJSON([1.5])).toThrow();
    expect(() => TextOperation.fromJSON("nope")).toThrow();
  });
});

describe("apply", () => {
  test("errors on length mismatch", () => {
    expect(() => op().retain(3).apply("ab")).toThrow(/length/);
  });
});

describe("invert", () => {
  test("undoes the operation", () => {
    const doc = "hello *world*";
    const o = op().retain(6).delete(7).insert("there");
    const applied = o.apply(doc);
    expect(o.invert(doc).apply(applied)).toBe(doc);
  });
});

describe("compose", () => {
  test("composition equals sequential application", () => {
    const doc = "abcdef";
    const a = op().retain(3).insert("X").retain(3);
    const b = op().delete(2).retain(5);
    expect(a.compose(b).apply(doc)).toBe(b.apply(a.apply(doc)));
  });

  test("rejects incompatible lengths", () => {
    expect(() => op().retain(3).compose(op().retain(4))).toThrow(/compose/);
  });
});

describe("transform", () => {
  test("TP1: both orders converge (Elixir fixture)", () => {
    const doc = "she is a girl";
    const a = op().retain(9).insert("good ").retain(4);
    const b = op().delete(3).insert("he").retain(10);

    const [a1, b1] = TextOperation.transform(a, b);
    expect(b1.apply(a.apply(doc))).toBe(a1.apply(b.apply(doc)));
    expect(b1.apply(a.apply(doc))).toBe("he is a good girl");
  });

  test("insert tie-break: left operand's insert comes first (Elixir fixture)", () => {
    const doc = "ab";
    const a = op().retain(1).insert("X").retain(1);
    const b = op().retain(1).insert("Y").retain(1);

    const [a1, b1] = TextOperation.transform(a, b);
    expect(b1.apply(a.apply(doc))).toBe("aXYb");
    expect(a1.apply(b.apply(doc))).toBe("aXYb");
  });

  test("overlapping deletes converge (Elixir fixture)", () => {
    const doc = "abcdef";
    const a = op().retain(1).delete(3).retain(2);
    const b = op().retain(2).delete(3).retain(1);

    const [a1, b1] = TextOperation.transform(a, b);
    expect(b1.apply(a.apply(doc))).toBe(a1.apply(b.apply(doc)));
    expect(b1.apply(a.apply(doc))).toBe("af");
  });

  test("rejects different base lengths", () => {
    expect(() => TextOperation.transform(op().retain(3), op().retain(4))).toThrow(/transform/);
  });
});

describe("transformIndex", () => {
  test("inserts before the index shift it right", () => {
    expect(TextOperation.transformIndex(3, op().insert("ab").retain(5))).toBe(5);
  });

  test("inserts after the index leave it alone", () => {
    expect(TextOperation.transformIndex(3, op().retain(4).insert("ab").retain(1))).toBe(3);
  });

  test("deletes spanning the index clamp it", () => {
    expect(TextOperation.transformIndex(3, op().retain(1).delete(4))).toBe(1);
  });

  test("deletes before the index shift it left", () => {
    expect(TextOperation.transformIndex(4, op().delete(2).retain(3))).toBe(2);
  });
});

describe("code point helpers", () => {
  test("cpLength counts code points", () => {
    expect(cpLength("a👍b")).toBe(3);
    expect("a👍b".length).toBe(4);
  });

  test("utf16 <-> code point conversions round-trip", () => {
    const doc = "a👍é🚀b"; // mixed BMP / astral / combining
    for (let cp = 0; cp <= cpLength(doc); cp++) {
      const u16 = cpToUtf16Offset(doc, cp);
      expect(utf16OffsetToCp(doc, u16)).toBe(cp);
    }
  });
});
